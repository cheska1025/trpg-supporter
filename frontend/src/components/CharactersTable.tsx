import type { Character } from '@/types/characters';

type Props = {
  items: Character[];
  onDelete?: (id: number) => void;
  onEdit?: (id: number, patch: Partial<Pick<Character, 'name' | 'clazz' | 'level'>>) => void;
};

export function CharactersTable({ items, onDelete, onEdit }: Props) {
  return (
    <table>
      <thead>
        <tr>
          <th>ID â–¾</th>
          <th>Name</th>
          <th>Class</th>
          <th>Level</th>
          {(onDelete || onEdit) && <th>Actions</th>}
        </tr>
      </thead>
      <tbody>
        {items.length === 0 ? (
          <tr>
            <td colSpan={5} style={{ textAlign: 'center' }}>
              No data
            </td>
          </tr>
        ) : (
          items.map((c) => (
            <tr key={c.id}>
              <td>{c.id}</td>
              <td>{c.name}</td>
              <td>{c.clazz}</td>
              <td>{c.level}</td>
              {(onDelete || onEdit) && (
                <td>
                  {onEdit && (
                    <button onClick={() => onEdit(c.id, { level: c.level + 1 })}>+1 Lv</button>
                  )}
                  {onDelete && <button onClick={() => onDelete(c.id)}>Delete</button>}
                </td>
              )}
            </tr>
          ))
        )}
      </tbody>
    </table>
  );
}
