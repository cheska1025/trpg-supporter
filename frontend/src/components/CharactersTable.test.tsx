import { describe, it, expect } from 'vitest';
import { render, within } from '@testing-library/react';
import { withClient } from '@/test/utils';
import { CharactersTable } from './CharactersTable';
import type { Character } from '@/types/characters';

describe('CharactersTable', () => {
  it('renders empty', () => {
    const { container, getByText } = render(
      withClient(<CharactersTable items={[]} onEdit={() => {}} onDelete={() => {}} />)
    );
    expect(getByText(/No data/i)).toBeInTheDocument();
  });

  it('renders some rows', () => {
    const items: Character[] = [{ id: 1, name: 'A', clazz: 'W', level: 1 }];
    const { container } = render(
      withClient(<CharactersTable items={items} onEdit={() => {}} onDelete={() => {}} />)
    );

    const tbody = container.querySelector('tbody')!;
    expect(within(tbody).getByText('A')).toBeInTheDocument();
    expect(within(tbody).getByText('W')).toBeInTheDocument();
  });
});
