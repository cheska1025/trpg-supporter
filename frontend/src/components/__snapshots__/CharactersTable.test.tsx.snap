// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharactersTable > renders empty 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          ID â–¾
        </th>
        <th>
          Name
        </th>
        <th>
          Class
        </th>
        <th>
          Level
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colspan="5"
          style="text-align: center;"
        >
          No data
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
